flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutor.java

```diff
@@ -35,27 +79,634 @@
+	private final IOManager ioManager;
+
+	/** The memory manager component in the task manager */
+	private final MemoryManager memoryManager;
+
+	/** The network component in the task manager */
+	private final NetworkEnvironment networkEnvironment;
+
+	/** The number of slots in the task manager, should be 1 for YARN */
+	private final int numberOfSlots;
 
+	// --------- resource manager --------
 	private TaskExecutorToResourceManagerConnection resourceManagerConnection;
 
 	// ------------------------------------------------------------------------
 
 	public TaskExecutor(
+			TaskManagerConfiguration taskManagerConfig,
+			ResourceID resourceID,
+			InstanceConnectionInfo connectionInfo,
+			MemoryManager memoryManager,
+			IOManager ioManager,
+			NetworkEnvironment networkEnvironment,
+			int numberOfSlots,
 			RpcService rpcService,
-			HighAvailabilityServices haServices,
-			ResourceID resourceID) {
+			HighAvailabilityServices haServices) {
 
 		super(rpcService);
 
-		this.haServices = checkNotNull(haServices);
+		this.taskManagerConfig = checkNotNull(taskManagerConfig);
 		this.resourceID = checkNotNull(resourceID);
+		this.connectionInfo = checkNotNull(connectionInfo);
+		this.memoryManager = checkNotNull(memoryManager);
+		this.ioManager = checkNotNull(ioManager);
+		this.networkEnvironment = checkNotNull(networkEnvironment);
+		this.numberOfSlots = checkNotNull(numberOfSlots);
+		this.haServices = checkNotNull(haServices);
+	}
+
+	/**
+	 * Starts and runs the TaskManager.
+	 * <p/>
+	 * This method first tries to select the network interface to use for the TaskManager
+	 * communication. The network interface is used both for the actor communication
+	 * (coordination) as well as for the data exchange between task managers. Unless
+	 * the hostname/interface is explicitly configured in the configuration, this
+	 * method will try out various interfaces and methods to connect to the JobManager
+	 * and select the one where the connection attempt is successful.
+	 * <p/>
+	 * After selecting the network interface, this method brings up an actor system
+	 * for the TaskManager and its actors, starts the TaskManager's services
+	 * (library cache, shuffle network stack, ...), and starts the TaskManager itself.
+	 *
+	 * @param configuration    The configuration for the TaskManager.
+	 * @param taskManagerClass The actor class to instantiate.
+	 *                         Allows to use TaskManager subclasses for example for YARN.
+	 */
+	public static void selectNetworkInterfaceAndRunTaskManager(
+		Configuration configuration,
+		ResourceID resourceID,
+		Class<? extends TaskManager> taskManagerClass) throws Exception {
+
+		Tuple2<String, Integer> tuple2 = selectNetworkInterfaceAndPort(configuration);
+
+		runTaskManager(tuple2._1(), resourceID, tuple2._2(), configuration, taskManagerClass);
+	}
+
+	private static Tuple2<String, Integer> selectNetworkInterfaceAndPort(Configuration configuration)
```

Comment:
Do we want to rely on Scala tuples here after the conversion from Scala to Java? I would suggest `InetAddres` as a return type here.